src/circuit.rs

```diff
@@ -41,6 +42,69 @@ impl From<JubJubAffine> for PublicInputValue {
     }
 }
 
+#[derive(Debug, Clone)]
+/// Collection of structs/objects that the Verifier will use in order to
+/// de/serialize data needed for Circuit proof verification.
+/// This structure can be seen as a link between the [`Circuit`] public input
+/// positions and the [`VerifierKey`] that the Verifier needs to use.
+pub struct VerifierData {
+    pub(self) key: VerifierKey,
+    pub(self) pi_pos: Vec<usize>,
+}
+
+impl VerifierData {
+    /// Creates a new [`VerifierData`] from a [`VerifierKey`] and the public
+    /// input positions of the circuit that it represents.
+    pub const fn new(key: VerifierKey, pi_pos: Vec<usize>) -> Self {
+        Self { key, pi_pos }
+    }
+
+    /// Returns a reference to the contained [`VerifierKey`].
+    pub const fn key(&self) -> &VerifierKey {
+        &self.key
+    }
+
+    /// Returns a reference to the contained Public Input positions.
+    pub const fn pi_pos(&self) -> &Vec<usize> {
+        &self.pi_pos
+    }
+
+    /// Deserializes the [`VerifierData`] into a vector of bytes.
+    #[allow(unused_must_use)]
+    pub fn to_var_bytes(&self) -> Vec<u8> {
+        let mut buff =
+            vec![
+                0u8;
+                VerifierKey::SIZE + u32::SIZE + self.pi_pos.len() * u32::SIZE
+            ];
+        let mut writer = &mut buff[..];
+
+        writer.write(&self.key.to_bytes());
+        writer.write(&(self.pi_pos.len() as u32).to_bytes());
+        self.pi_pos.iter().copied().for_each(|pos| {
+            // Omit the result since disk_bytes write can't fail here
+            // due to the fact that we're writing into a vector basically.
+            let _ = writer.write(&(pos as u32).to_bytes());
+        });
+
+        buff
+    }
+
+    /// Serializes [`VerifierData`] from a slice of bytes.
+    pub fn from_slice(buf: &[u8]) -> Result<Self, Error> {
```

Comment:
I would not constrain that in the signature. Maybe the user can't provide a mut ref. And we just need it mutable because of reader reasons (internal to the fn) not because we want to change the original slice.

Indeed we want to leave the original `buf` intact.

Reply:
The mutability on the proposed line is restricted to the scope of the function and doesn't change the signature. For both cases, the signature is `fn(&[u8]) -> Result<Self, Error>`

Reply:
Ohh I missread it. Agree. I thought you purposed `&mut [u8]`